modules/tinymce/src/plugins/anchor/main/ts/core/Anchor.ts

```diff
@@ -5,37 +5,88 @@
  * For commercial licenses see https://www.tiny.cloud/
  */
 
+import { Node, HTMLAnchorElement } from '@ephox/dom-globals';
+import { Arr, Option } from '@ephox/katamari';
 import Editor from 'tinymce/core/api/Editor';
-import { Element } from '@ephox/dom-globals';
+import RangeUtils from 'tinymce/core/api/dom/RangeUtils';
+import * as Settings from '../api/Settings';
 
-const isNamedAnchor = (editor: Editor, node: Element) => node.tagName === 'A' && editor.dom.getAttrib(node, 'href') === '';
+const namedAnchorSelector = 'a:not([href])';
+
+const isAnchor = (elm: Node): elm is HTMLAnchorElement => elm && elm.nodeName.toLowerCase() === 'a';
+const isNamedAnchor = (elm: Node): elm is HTMLAnchorElement => isAnchor(elm) && !elm.getAttribute('href') && (!!elm.getAttribute('id') || !!elm.getAttribute('name'));
+const isEmptyNamedAnchor = (elm: Node): elm is HTMLAnchorElement => isNamedAnchor(elm) && !elm.firstChild;
+
+const removeEmptyNamedAnchorsInSelection = (editor: Editor) => {
+  const dom = editor.dom;
+  RangeUtils(dom).walk(editor.selection.getRng(), (nodes) => {
+    Arr.each(nodes, (node) => {
+      if (isEmptyNamedAnchor(node)) {
+        dom.remove(node, false);
+      }
+    });
+  });
+};
 
 const isValidId = (id: string) =>
   // Follows HTML4 rules: https://www.w3.org/TR/html401/types.html#type-id
   /^[A-Za-z][A-Za-z0-9\-:._]*$/.test(id);
 
-const getId = (editor: Editor) => {
-  const selectedNode = editor.selection.getNode();
-  return isNamedAnchor(editor, selectedNode) ? (selectedNode.getAttribute('id') || selectedNode.getAttribute('name')) : '';
-};
+const getNamedAnchor = (editor: Editor): Option<HTMLAnchorElement> => Option.from(editor.dom.getParent(editor.selection.getStart(), namedAnchorSelector) as HTMLAnchorElement);
 
-const insert = (editor: Editor, id: string) => {
-  const selectedNode = editor.selection.getNode();
+const getId = (editor: Editor) => getNamedAnchor(editor).map((anchor) => anchor.getAttribute('id') || anchor.getAttribute('name')).getOr('');
 
-  if (isNamedAnchor(editor, selectedNode)) {
-    selectedNode.removeAttribute('name');
-    selectedNode.id = id;
-    editor.undoManager.add();
-  } else {
-    editor.focus();
+const createAnchor = (editor: Editor, id: string): void => {
+  editor.focus();
```

Comment:
This is more of a legacy bug... however it seems very strange to focus the editor on insert, but not on update. This should probably be pulled out to apply to both cases.